core/trino-main/src/main/java/io/trino/operator/output/PositionsAppenderPageBuilder.java

```diff
@@ -0,0 +1,129 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.operator.output;
+
+import io.trino.spi.Page;
+import io.trino.spi.block.Block;
+import io.trino.spi.block.PageBuilderStatus;
+import io.trino.spi.type.Type;
+
+import java.util.List;
+
+import static com.google.common.base.Preconditions.checkState;
+import static io.trino.spi.PageBuilder.DEFAULT_INITIAL_EXPECTED_ENTRIES;
+import static java.util.Objects.requireNonNull;
+
+public class PositionsAppenderPageBuilder
+{
+    private final PositionsAppender[] positionsAppenders;
+    private final List<? extends Type> types;
+    private final PositionsAppenderFactory positionsAppenderFactory;
+    private PageBuilderStatus pageBuilderStatus;
```

Comment:
we don't need `PageBuilderStatus` mechanics to decide if we need to construct output page. `PageBuilderStatus` is relatively simple, it just checks current page size against `DEFAULT_MAX_PAGE_SIZE_IN_BYTES` (by default).

Reply:
I have a feeling that you have a strong desire to keep this backward-compatible with the existing code, while me and @sopel39 would like to explicitly make it non-compatible not to carry any burden of compatibility